pkg/cloudprovider/provider/azure/provider.go

```diff
@@ -0,0 +1,581 @@
+		Offer:     to.StringPtr("Canonical"),
+		Sku:       to.StringPtr("18.04-LTS"),
+		Version:   to.StringPtr("latest"),
+	},
+}
+
+func getOSImageReference(os providerconfig.OperatingSystem) (*compute.ImageReference, error) {
+	ref, supported := imageReferences[os]
+	if !supported {
+		return nil, fmt.Errorf("operating system %q not supported", os)
+	}
+
+	return &ref, nil
+}
+
+// New returns a digitalocean provider
+func New(configVarResolver *providerconfig.ConfigVarResolver) cloud.Provider {
+	return &provider{configVarResolver: configVarResolver}
+}
+
+func (p *provider) getConfig(s runtime.RawExtension) (*config, *providerconfig.Config, error) {
+	pconfig := providerconfig.Config{}
+	err := json.Unmarshal(s.Raw, &pconfig)
+	if err != nil {
+		return nil, nil, err
+	}
+	rawCfg := rawConfig{}
+	err = json.Unmarshal(pconfig.CloudProviderSpec.Raw, &rawCfg)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c := config{}
+	c.SubscriptionID, err = p.configVarResolver.GetConfigVarStringValue(rawCfg.SubscriptionID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"subscriptionID\" field, error = %v", err)
+	}
+
+	c.TenantID, err = p.configVarResolver.GetConfigVarStringValue(rawCfg.TenantID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"tenantID\" field, error = %v", err)
+	}
+
+	c.ClientID, err = p.configVarResolver.GetConfigVarStringValue(rawCfg.ClientID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"clientID\" field, error = %v", err)
+	}
+
+	c.ClientSecret, err = p.configVarResolver.GetConfigVarStringValue(rawCfg.ClientSecret)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"clientSecret\" field, error = %v", err)
+	}
+
+	c.ResourceGroup, err = p.configVarResolver.GetConfigVarStringValue(rawCfg.ResourceGroup)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"resourceGroup\" field, error = %v", err)
+	}
+
+	c.Location, err = p.configVarResolver.GetConfigVarStringValue(rawCfg.Location)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"location\" field, error = %v", err)
+	}
+
+	c.VMSize, err = p.configVarResolver.GetConfigVarStringValue(rawCfg.VMSize)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"vmSize\" field, error = %v", err)
+	}
+
+	c.VNetName, err = p.configVarResolver.GetConfigVarStringValue(rawCfg.VNetName)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"vnetName\" field, error = %v", err)
+	}
+
+	c.SubnetName, err = p.configVarResolver.GetConfigVarStringValue(rawCfg.SubnetName)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"subnetName\" field, error = %v", err)
+	}
+
+	return &c, &pconfig, nil
+}
+
+func getVMIPAddresses(ctx context.Context, c *config, vm *compute.VirtualMachine) ([]string, error) {
+	ipAddresses := make([]string, 0)
```

Comment:
I think its more idiomatic to write `var ipAddresses []string` instead of instantiating a zero-lenght array

Reply:
fixed